hoot-core/src/main/cpp/hoot/core/io/ApiDbReader.cpp

```diff
@@ -402,19 +404,110 @@ void ApiDbReader::_readByBounds(OsmMapPtr map, const Envelope& bounds)
         }
       }
       LOG_VARD(relationIds.size());
-
-      if (relationIds.size() > 0)
+      //  Iterate all relations (and sub-relations) that are "within" the bounds
+      QSet<QString> completedRelationIds;
+      while (relationIds.size() > 0)
       {
+        QSet<QString> newNodes;
+        QSet<QString> newWays;
+        QSet<QString> newRelations;
         LOG_DEBUG("Retrieving relations by relation ID...");
         boost::shared_ptr<QSqlQuery> relationItr =
           _getDatabase()->selectElementsByElementIdList(relationIds, TableType::Relation);
         while (relationItr->next())
         {
           RelationPtr relation = _resultToRelation(*relationItr, *map);
           map->addElement(relation);
+          const vector<RelationData::Entry>& members = relation->getMembers();
+          //  Iterate all members so that they can be retrieved later
+          for (vector<RelationData::Entry>::const_iterator it = members.begin(); it != members.end(); ++it)
+          {
+            ElementType type = it->getElementId().getType();
+            QString id = QString::number(it->getElementId().getId());
+            if (type == ElementType::Node)
+              newNodes.insert(id);
+            else if (type == ElementType::Way)
+              newWays.insert(id);
+            else if (type == ElementType::Relation)
+              newRelations.insert(id);
+            else
+            {
+              LOG_WARN("Unknown relation member type");
+            }
+          }
           LOG_VART(relation);
           boundedRelationCount++;
+          //  Keep track of all relations that have been added to the map so we don't try them again
+          completedRelationIds.insert(QString::number(relation->getId()));
+        }
+        //  Clear the relations that we are iterating on, it is then filled with new relations later
+        relationIds.clear();
+        //  Iterate any new nodes that are members of relations that need to be queried
+        newNodes = newNodes.subtract(nodeIds);
+        if (newNodes.size() > 0)
+        {
+          boost::shared_ptr<QSqlQuery> nodeItr = _getDatabase()->selectElementsByElementIdList(newNodes, TableType::Node);
+          while (nodeItr->next())
+          {
+            const QSqlQuery resultIterator = *nodeItr;
+            NodePtr node = _resultToNode(resultIterator, *map);
+            LOG_VART(node);
+            map->addElement(node);
+            boundedNodeCount++;
+            const long nodeId = resultIterator.value(0).toLongLong();
+            LOG_VART(ElementId(ElementType::Node, nodeId));
+            nodeIds.insert( QString::number(nodeId));
+          }
+        }
+        //  Iterate any new ways that are members of relations that need to be queried
```

Comment:
Query any nodes that are members of relations that haven't been pulled down yet.